src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs

```diff
@@ -0,0 +1,159 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed partial class EditorConfig
+    {
+        /// <summary>
+        /// Takes a <see cref="Section.Name"/> and compiles the file glob
+        /// inside to a regex which recognizes the given language.
+        /// </summary>
+        public static string CompileSectionNameToRegEx(string sectionName)
+        {
+            // An editorconfig section name is a language for recognizing file paths
+            // defined by the following grammar:
+            //
+            // <path> ::= <path-list>
+            // <path-list> ::= <path-item> | <path-item> <path-list>
+            // <path-item> ::= "*"  | "**" | "?" | <char> | <choice> | <range>
+            // <char> ::= any unicode character
+            //
+            // PROTOTYPE(editorconfig): Below remains unimplemented
+            //
+            // <choice> ::= "{" <choice-list> "}" 
+            // <choice-list> ::= <path-list> | <path-list> "," <choice-list>
+            // <range> ::= "{" <integer> ".." <integer> "}"
+            // <integer> ::= "-" <digit-list> | <digit-list>
+            // <digit-list> ::= <digit> | <digit> <digit-list>
+            // <digit> ::= 0-9
+
+            var sb = new StringBuilder();
+            sb.Append('^');
+            var lexer = new SectionNameLexer(sectionName);
+            while (!lexer.IsDone)
+            {
+                var tokenKind = lexer.Lex();
+                switch (tokenKind)
+                {
+                    case TokenKind.BadToken:
+                        // Parsing failure
+                        return null;
+                    case TokenKind.SimpleCharacter:
+                        // Matches just this character
+                        sb.Append(lexer.EatCurrentCharacter());
+                        break;
+                    case TokenKind.Question:
+                        // '?' matches any single character
+                        sb.Append('.');
+                        break;
+                    case TokenKind.Star:
+                        // Matches any string of characters except directory separator
+                        // Directory separator is defined in editorconfig spec as '/'
```

Comment:
>// Directory separator is defined in editorconfig spec as '/' [](start = 24, length = 61)

I assume that windows developers will use \ irrespective of what the spec says. Where do we normalize the values or error out on bad slashes?